lib/queryByExample.js

```diff
@@ -0,0 +1,232 @@
+    // no example to match against, return all credentials
+    return credentials;
+  }
+
+  // Convert example to JSON pointers, excluding @context as it's handled
+  // separately
+  const expectedPointers = _convertExampleToPointers(example);
+
+  if(expectedPointers.length === 0) {
+    // no meaningful fields to match, return all credentials
+    return credentials;
+  }
+
+  return credentials.filter(credential => {
+    // Check each pointer against the credential content
+    return expectedPointers.every(({pointer, expectedValue}) => {
+      try {
+        const actualValue = JsonPointer.get(credential, pointer);
+        const result = _valuesMatch(actualValue, expectedValue);
+
+        return result;
+      } catch(e) {
+        // If pointer doesn't exist in credential, it's not a match
+        console.log('Pointer error:', pointer, e.message);
+        return false;
+      }
+    });
+  });
+}
+
+/**
+ * Converts an example to an array of JSON pointer/value pairs.
+ * This function recursively processes the example object to extract all
+ * field paths and their expected values, excluding @context which is
+ * handled separately in the filtering pipeline.
+ *
+ * @param {object} example - The example object from Query By Example.
+ *
+ * @returns {Array<object>} Array of objects with {pointer, expectedValue}
+ *  where pointer is a JSON pointer string (e.g., '/credentialSubject/name')
+ *  and expectedValue is the expected value at the path.
+ */
+function _convertExampleToPointers(example) {
+  const pointers = [];
+
+  // Create a copy without @context since it's handled by _matchContextFilter
+  const exampleWithoutContext = {...example};
+  delete exampleWithoutContext['@context'];
+
+  // Convert to JSON pointer dictionary and extract pointer/value pairs
+  try {
+    const dict = JsonPointer.dict(exampleWithoutContext);
+    for(const [pointer, value] of Object.entries(dict)) {
+      // Skip empty objects, arrays, or null/undefined values
+      if(_isMatchableValue(value)) {
+        pointers.push({
+          pointer,
+          expectedValue: value
+        });
+      }
+    }
+  } catch(e) {
+    // If JSON pointer conversion fails, return empty array
+    console.warn('Failed to convert example to JSON pointers:', e);
+    return [];
+  }
+  return pointers;
+}
+
+/**
+ * Determines if a value is suitable for matching. We skip empty objects,
+ * empty arrays, null, undefined, and other non-meaningful values.
+ *
+ * @param {*} value - The value to check.
+ *
+ * @returns {boolean} True if the value should be used for matching.
+ */
+function _isMatchableValue(value) {
+  // Skip null, undefined
+  if(value == null) {
+    return false;
+  }
+
+  // Skip empty arrays
+  if(Array.isArray(value) && value.length === 0) {
+    return false;
+  }
+
+  // Skip empty objects
+  if(typeof value === 'object' && !Array.isArray(value) &&
+      Object.keys(value).length === 0) {
+    return false;
+  }
```

Comment:
I think empty objects are considered "wildcards" -- meaning, any value will match. Clearly under-specified in the spec, but I think this was the aim.